File: CHUONG_7/7.09_MultiThread/app/src/main/java/com/example/myapplication/repository/LoadSongRepository.java
package com.example.myapplication.repository;

import com.example.myapplication.model.Song;
import com.example.myapplication.model.SongList;
import com.example.myapplication.resource.Result;
import com.example.myapplication.resource.SongResponseParser;

import java.io.IOException;
import java.util.concurrent.Executor;

public class LoadSongRepository {
    private final Executor executor;
    private final SongResponseParser parser;

    public LoadSongRepository(SongResponseParser parser, Executor executor) {
        this.executor = executor;
        this.parser = parser;
    }

    public void makeGetSongRequest(final RepositoryCallback<SongList> callback){
        executor.execute(() -> {
            try {
                Result<SongList> result = makeSyncGetSongRequest();
                callback.onComplete(result);
            } catch (Exception ex) {
                Result<SongList> errorResult = new Result.Error<>(ex);
                callback.onComplete(errorResult);
            }
        });
    }

    public Result<SongList> makeSyncGetSongRequest() {
        String url = "https://thantrieu.com/resources/braniumapis/song.json";
        try {
            SongList songList = parser.parse(url);
            return new Result.Success<>(songList);
        } catch (IOException ex) {
            return new Result.Error<>(ex);
        }
    }


}

File: CHUONG_7/7.08_Coroutine/app/src/main/java/com/example/coroutine/repository/LoadSongRepository.kt
package com.example.coroutine.repository

import com.example.coroutine.model.SongList
import com.example.coroutine.resource.SongResponseParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.IOException
import com.example.coroutine.resource.Result

class LoadSongRepository (
    private val parser: SongResponseParser
) {
    companion object {
        const val URL = "https://thantrieu.com/resources/braniumapis/song.json"
    }
    suspend fun loadSongs(): Result<SongList> {
        return withContext(Dispatchers.IO) {
            val result = try {
                val response = parser.parse(URL)
                Result.Success(response)
            } catch (ex: IOException) {
                Result.Error(ex)
            }
            result
        }
    }
}